Sources/Credentials/CredentialsTokenTTL.swift

```diff
@@ -0,0 +1,147 @@
+/**
+* Copyright IBM Corporation 2016, 2017
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+**/
+
+import Kitura
+import KituraNet
+import Foundation
+
+/// Protocol to make it easier to add token TTL to credentials plugins.
+/// Using this protocol:
+/// Step 1) Conform to the protocol
+/// Step 2) Call the getProfileAndCacheIfNeeded method-- probably at the end of your authenticate method.
+public protocol CredentialsTokenTTL: AnyObject {
+    var usersCache: NSCache<NSString, BaseCacheElement>? {get}
+    var tokenTimeToLive: TimeInterval? {get}
+    
+    /// Used by the getProfileAndCacheIfNeeded method to generate a profile if one can't be used from cache.
+    /// - Parameter token: The Oauth2 token, used as a key in the cache.
+    /// - Parameter options: The dictionary of plugin specific options.
+    func generateNewProfile(token: String, options: [String:Any], completion: @escaping (CredentialsTokenTTLResult) -> Void)
+}
+
+public enum CredentialsTokenTTLResult {
+    // These two cases should be sufficient for most plugins. And are used by the getProfileAndCacheIfNeeded method with onSuccess and onFailure callbacks.
+    case success(UserProfile)
+    case failure(HTTPStatusCode?, [String:String]?)
+    
+    // This case is only used by the getProfileAndCacheIfNeeded method with the completion callback and is intended for plugins with more complicated needs.
+    case other(details: String)
```

Comment:
How about `unprocessable` rather than `other`? Also please can you add documentation to each of the public members, eg:
```swift
/// Represents the result of a call to `generateNewProfile()` with an authentication token.
/// On success, the resulting `UserProfile` is returned. On failure, the plugin may return
/// a status code and headers that should be sent in response. If the plugin cannot
/// process the token provided, then it may choose whether to fail, or to pass to allow
/// other plugins to handle authentication instead.
public enum CredentialsTokenTTLResult {
    /// Authentication was successful. The `UserProfile` represents the identity of the bearer.
    case success(UserProfile)
    /// The token was successfully parsed, but authentication failed. The plugin may provide
    /// a status code and headers to send in response.
    case failure(HTTPStatusCode?, [String:String]?)
    /// The token could not be handled by this plugin. It may be malformed, or intended for
    /// another plugin.
    case unprocessable(details: String)
```